mrmustard/lab/abstract/state.py

```diff
@@ -390,11 +396,35 @@ def primal(self, other: Union[State, Transformation]) -> State:
 
     def __and__(self, other: State) -> State:
         r"""Concatenates two states."""
-        if not (self.is_gaussian and other.is_gaussian):
-            raise NotImplementedError(
-                "Concatenation of non-gaussian states is not implemented yet."
-            )
-
+        if not self.is_gaussian or not other.is_gaussian:  # convert all to fock now
+            if (
+                self.is_mixed or other.is_mixed
+            ):  # TODO: would be more efficient if we could keep pure states as kets
```

Comment:
(minor cosmetic suggestion)
```suggestion
            # TODO: would be more efficient if we could keep pure states as kets
            if self.is_mixed or other.is_mixed:
```